### PR TITLE
[Feat]: Word-Level Navigation for Text-Heavy HTML in Live Preview

### LINKED ISSUE #2239: [Feat]: Word-Level Navigation for Text-Heavy HTML in Live Preview
## Customer Request:
"In Live Preview, I love that you can click a section and it goes to the corresponding code. The other day, I needed to find where a word occurred on a page with a lot of text, but it only went to the section. So the ability to go to a specific word would be appreciated."

## Proposed Feature:
Enhanced Live Preview navigation with word-level precision for text-heavy HTML

## Benefits:
- More precise navigation between preview and code
- Ability to locate specific words or phrases in text-heavy HTML documents
- Time-saving when working with content-rich pages
- Improved editing workflow for HTML content creators

## Implementation:
- Extend the existing section-level navigation to work at word level
- Add highlighting for the specific word in both preview and code views
- Maintain compatibility with the current section navigation system
- Consider keyboard shortcuts for toggling between section and word navigation modes

### PR DESCRIPTION
**Pull request by atondapu**
PR Type: Feature

https://github.com/phcode-dev/phoenix/issues/2239

Fixes #2239

Feature Description
This PR adds the ability to navigate directly to individual words in Live Preview rather than just sections. When enabled, clicking on a word in the Live Preview will position the editor cursor at the exact character offset of that word in the source code, making it easier to find and edit specific content in text-heavy documents.

Screenshots: Find attached

Screenshot 2025-04-25 at 5 41 36 PM

Breaking Change: No

Tests Done: Passes all listing, existing tests, underwent manual general and edge-case testing to ensure correct word is found when button is toggled. (works across different browsers, and OS as before)

Additional Information
The implementation extends the existing section-level click handler in Live Preview to calculate the exact word position. A new toggle button with a translucent document glyph allows users to switch between section-level and word-level navigation modes. Special care was taken to handle edge cases with duplicate words by using click coordinates to identify the precise text node and word instance.

### CODE REVIEW

**Comment by abose** [comment]
Thanks for the contributing @atondapu 

To clarify the licensing situation, we'd like to request your explicit permission to distribute the changes you've submitted in this pull request.

Can you do one of the following to help me continue with this change?

1. Sign the CLA at https://phcode-dev.github.io/contributor-license-agreement/ OR
1. Just comment,  "I grant permission to distribute my contributions in this pull request under [MIT License Terms](https://opensource.org/license/mit)."

File: src/extensionsIntegrated/Phoenix-live-preview/panel.html [hunk 1]
```diff
@@ -3,6 +3,7 @@
         <div style="width: 20%;display: flex;">
             <button id="reloadLivePreviewButton" title="{{clickToReload}}" class="btn-alt-quiet toolbar-button reload-icon"></button>
             <button id="highlightLPButton" title="{{toggleLiveHighlight}}" class="btn-alt-quiet toolbar-button pointer-fill-icon"></button>
+            <button id="wordNavigationLPButton" title="{{toggleWordNavigation}}" class="btn-alt-quiet toolbar-button text-icon"></button>
```
Comment (abose) [inline src/extensionsIntegrated/Phoenix-live-preview/panel.html:6]: We tend to avoid placing new buttons on the live preview unless its absolutely required to keep the UI clutter in check.

Since this can be the default behavior, please tie this to the already existing `toggleLiveHighlight` flag and remove the new button.

File: src/extensionsIntegrated/Phoenix-live-preview/main.js [hunk 2]
```diff
@@ -389,6 +411,7 @@ define(function (require, exports, module) {
         $iframe = $panel.find("#panel-live-preview-frame");
         $pinUrlBtn = $panel.find("#pinURLButton");
         $highlightBtn = $panel.find("#highlightLPButton");
+        $wordNavigationBtn = $panel.find("#wordNavigationLPButton");
```
Comment (abose) [inline src/extensionsIntegrated/Phoenix-live-preview/main.js:414]: We don't need a new button for this, we can remove all these code as it can be tied to `toggleLiveHighlight` button.

File: src/base-config/keyboard.json [hunk 3]
```diff
@@ -31,6 +31,9 @@
     "file.previewHighlight":  [
         "Ctrl-Shift-C"
     ],
+    "file.previewWordNavigation":  [
```
Comment (abose) [inline src/base-config/keyboard.json:34]: We usually add shortcuts only if it is very frequently used. since people are not likely to toggle this feature, a shortcut isn't necessary.

File: src/LiveDevelopment/main.js [hunk 4]
```diff
@@ -302,13 +315,22 @@ define(function main(require, exports, module) {
             _updateHighlightCheckmark();
         });
 
+    PreferencesManager.stateManager.definePreference("livedevWordNavigation", "boolean", false)
```
Comment (abose) [inline src/LiveDevelopment/main.js:318]: reuse the `livedevHighlight` button that already exists for this flow.

File: src/LiveDevelopment/MultiBrowserImpl/protocol/LiveDevProtocol.js [hunk 5]
```diff
@@ -147,31 +147,335 @@ define(function (require, exports, module) {
         }
     }
 
-    function _tagSelectedInLivePreview(tagId, nodeName, contentEditable, allSelectors) {
+    function _findWordInEditor(editor, word, nodeName, tagId, wordContext, wordOccurrenceIndex) {
```
Comment (abose) [inline src/LiveDevelopment/MultiBrowserImpl/protocol/LiveDevProtocol.js:150]: Can you give a high level overview of how this works in plain text? It will help me understand the changes. Specifically:
1. How we identify the text in the remote live preview.
2. How it communicates this position and finds the correct text position in the editor.

**Review by abose — COMMENTED** [review]
Thanks for this feature! I have left a few comments mainly to help me understand how this is implemented for reviewing.

 Also how was this tested(In desktop app/browser)? Did you run the test suite?

File: src/LiveDevelopment/BrowserScripts/LiveDevProtocolRemote.js [hunk 6]
```diff
@@ -382,12 +382,146 @@
...
+                }
+            }
+
+            return null;
+        }
+
+        const nodeText = textNode.textContent;
+
+        // Function to extract a word and its occurrence index
+        function extractWordAndOccurrence(text, wordStart, wordEnd) {
```
Comment (abose) [inline src/LiveDevelopment/BrowserScripts/LiveDevProtocolRemote.js:429]: what will happen if we don't do word search? is the text not not always the correct target on click?

File: src/LiveDevelopment/MultiBrowserImpl/protocol/LiveDevProtocol.js [hunk 7]
```diff
@@ -147,31 +147,335 @@ define(function (require, exports, module) {
...
+                    }
+
+                    if (found) {
+                        return true;
+                    }
+
+                }
+
+                // If occurrence index search failed or no occurrence index available, try context as a fallback
+                if (wordContext) {
```
Comment (abose) [inline src/LiveDevelopment/MultiBrowserImpl/protocol/LiveDevProtocol.js:248]: i see that there are a few fallbacks here, What cases triggers these fullbacks? A few examples can help me understand the edge cases better. We may also need to add live preview integration tests for those cases later.

File: src/LiveDevelopment/BrowserScripts/LiveDevProtocolRemote.js [hunk 8]
```diff
@@ -382,12 +382,146 @@
...
+    /**
+    * Gets the word at the clicked position along with additional information
+    * @param {Element} element - The element that was clicked
+    * @param {MouseEvent} event - The click event
+    * @return {Object|null} - Object containing the word and additional info, or null if not found
+    */
+    function getClickedWord(element, event) {
+
+        // Try to find the clicked position within the element
+        const range = document.caretRangeFromPoint(event.clientX, event.clientY);
```
Comment (abose) [inline src/LiveDevelopment/BrowserScripts/LiveDevProtocolRemote.js:394]: This api doesn't work in Firefox, so maybe we can fallback to [caretPositionFromPoint](https://developer.mozilla.org/en-US/docs/Web/API/Document/caretPositionFromPoint) in Firefox(caretPositionFromPoint doesnt work in safari), or disable the feature in Firefox.

**Comment by abose** [comment]
@atondapu This is a cool feature to have. Would you be able to check this out or need any assistance?

**Review by bartzjegr — APPROVED** [review]
approve
